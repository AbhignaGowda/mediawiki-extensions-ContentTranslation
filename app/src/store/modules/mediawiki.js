import pageApi from "../../wiki/mw/api/page";
import siteApi from "../../wiki/mw/api/site";

const state = {
  pages: [],
  languageInfo: {},
  languageTitleGroups: [],
  supportedLanguageCodes: []
};

const mutations = {
  addPage(state, page) {
    state.pages.push(page);
  },
  addLanguageTitleGroup(state, group) {
    state.languageTitleGroups.push(group);
  },
  setLanguageInfo(state, languageInfo) {
    state.languageInfo = languageInfo;
  },
  setSupportedLanguageCodes(state, languageCodes) {
    state.supportedLanguageCodes = languageCodes;
  }
};

// Computed properties for stores.
const getters = {
  getPage: state => (language, title) =>
    state.pages.find(
      page =>
        page.language === language &&
        (page.title === title || page.alias === title)
    ),
  getLanguageTitleGroup: state => (language, title) =>
    state.languageTitleGroups.find(group =>
      group.titles.find(
        groupTitle => groupTitle.lang === language && groupTitle.title === title
      )
    ),
  getLanguageTitleGroupByWikidataId: state => wikidataId =>
    state.languageTitleGroups.find(group => group.wikidataId === wikidataId),
  titleExistsInLanguageForGroup: (state, getters) => (wikidataId, language) =>
    (getters.getLanguageTitleGroupByWikidataId(wikidataId)?.titles || []).some(
      title => title.lang === language
    ),
  getTitleByLanguageForGroup: (state, getters) => (wikidataId, language) =>
    (getters.getLanguageTitleGroupByWikidataId(wikidataId)?.titles || []).find(
      title => title.lang === language
    )?.title
};

const actions = {
  fetchPageMetadata({ getters, commit }, request) {
    const titles = request.titles.filter(
      title => !getters.getPage(request.language, title)
    );

    const chunkSize = 50;
    for (let i = 0; i < titles.length; i += chunkSize) {
      let titlesSubset = titles.slice(i, i + chunkSize);
      pageApi
        .fetchPages(request.language, titlesSubset)
        .then(metadataList =>
          metadataList.forEach(page => commit("addPage", page))
        );

      titlesSubset = titlesSubset.filter(
        title => !getters.getLanguageTitleGroup(request.language, title)
      );
      pageApi
        .fetchTitles(request.language, titlesSubset)
        .then(groupList =>
          groupList.forEach(group => commit("addLanguageTitleGroup", group))
        );
    }
  },
  fetchLanguageInfo({ commit }) {
    siteApi.fetchLanguageInfo().then(languageInfo => {
      commit("setLanguageInfo", languageInfo);
    });
  },
  fetchSupportedLanguageCodes({ commit }) {
    siteApi.fetchSupportedLanguageCodes().then(languageCodes => {
      commit("setSupportedLanguageCodes", languageCodes);
    });
  }
};

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations
};
