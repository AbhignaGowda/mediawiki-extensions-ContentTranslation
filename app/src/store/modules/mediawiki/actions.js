import pageApi from "../../../wiki/mw/api/page";
import siteApi from "../../../wiki/mw/api/site";
import translatorApi from "../../../wiki/cx/api/translator";

export default {
  /**
   * @param {String} language
   * @param {Array<String>} titles
   * @returns {Promise<void>}
   */
  fetchPageMetadata({ getters, commit }, { language, titles }) {
    titles = titles.filter(title => !getters.getPage(language, title));

    const chunkSize = 50;

    for (let i = 0; i < titles.length; i += chunkSize) {
      let titlesSubset = titles.slice(i, i + chunkSize);
      // Avoid async/await here, so that requests can be sent in parallel
      pageApi.fetchPages(language, titlesSubset).then(metadataList => {
        metadataList.forEach(page => {
          commit("addPage", page);
        });
      });
    }
  },

  fetchLanguageTitles({ commit, getters }, { language, title }) {
    if (getters.getLanguageTitleGroup(language, title)) {
      // Already exist in store.
      return;
    }
    pageApi
      .fetchLanguageTitles(language, title)
      .then(
        languageTitleGroup =>
          languageTitleGroup &&
          commit("addLanguageTitleGroup", languageTitleGroup)
      );
  },

  fetchLanguages({ commit, state }) {
    // If languages have already been fetched, then skip
    if (!state.languages.length && !state.languagesRequested) {
      commit("setLanguagesRequested", true);
      siteApi.fetchLanguages().then(languages => {
        commit("setLanguages", languages);
      });
    }
  },

  fetchSupportedLanguageCodes({ commit, state }) {
    // If supported language codes have already been fetched, then skip
    if (
      !state.supportedLanguageCodes.length &&
      !state.supportedLanguageCodesRequested
    ) {
      commit("setSupportedLanguageCodesRequested", true);
      siteApi.fetchSupportedLanguageCodes().then(languageCodes => {
        commit("setSupportedLanguageCodes", languageCodes);
      });
    }
  },

  async fetchPageContent(
    { commit, getters, dispatch },
    { sourceLanguage, targetLanguage, sourceTitle }
  ) {
    let page = getters.getPage(sourceLanguage, sourceTitle);

    if (!page) {
      page = await pageApi.fetchPageContent(
        sourceLanguage,
        targetLanguage,
        sourceTitle
      );
      commit("addPage", page);
    }

    if (page.content) {
      return;
    }

    pageApi
      .fetchPageContent(sourceLanguage, targetLanguage, sourceTitle)
      .then(
        /** @type Page */ responsePage => (page.content = responsePage.content)
      );
  },

  /**
   * Returns a promise so that it can be awaited for
   * @param getters
   * @param commit
   * @param sourceLanguage
   * @param targetLanguage
   * @param sourceTitle
   * @return {Promise<PageSection[]>}
   */
  async fetchPageSections(
    { getters, commit },
    { sourceLanguage, targetLanguage, sourceTitle }
  ) {
    const page = getters.getPage(sourceLanguage, sourceTitle);

    if (!page) {
      return;
    }

    return pageApi
      .fetchPageSections(sourceLanguage, targetLanguage, sourceTitle)
      .then(sections => commit("setPageSections", { page, sections }));
  },

  /**
   * For a given language pair, this method checks for already existing
   * MT providers in the store, and if none, it fetches supported MT
   * providers and saves them to the store.
   * @param commit
   * @param getters
   * @param sourceLanguage
   * @param targetLanguage
   * @return {Promise<void>}
   */
  async fetchMTProviders(
    { commit, getters },
    { sourceLanguage, targetLanguage }
  ) {
    if (
      getters.getSupportedMTProviders(sourceLanguage, targetLanguage).length
    ) {
      return;
    }

    const mtProviderGroup = await siteApi.fetchSupportedMTProviders(
      sourceLanguage,
      targetLanguage
    );

    commit("addMtProviderGroup", mtProviderGroup);
  },

  /**
   * Translates HTML content for a given language pair
   * and MT provider, and returns a promise that resolves
   * to a string containing the translation.
   * @param getters
   * @param sourceLanguage
   * @param targetLanguage
   * @param provider
   * @param originalContent
   * @return {Promise<String>}
   */
  async translateSegment(
    { getters, rootGetters, dispatch },
    { sourceLanguage, targetLanguage, provider, originalContent }
  ) {
    const isValidProvider = getters.isValidProviderForTranslation(
      sourceLanguage,
      targetLanguage,
      provider
    );

    if (!isValidProvider) {
      return;
    }

    try {
      const token = await dispatch(
        "application/getCXServerToken",
        {},
        { root: true }
      );

      return await translatorApi.fetchSegmentTranslation(
        sourceLanguage,
        targetLanguage,
        provider,
        originalContent,
        token
      );
    } catch (error) {
      // Fall back to original content
      return originalContent;
    }
  },

  /**
   * Fetch nearby suggestions for current source language
   * based on user location, and store them to state, so that they
   * can be reused when "Search for an article" screen is mounted again.
   * If such suggestions already exist, the actions returns without doing anything.
   *
   * @param {Object} context
   * @param {Function} context.commit
   * @param {Object} context.rootState
   * @return {Promise<void>}
   */
  async fetchNearbyPages({ commit, rootState, state }) {
    const { sourceLanguage } = rootState.application;

    if (state.nearbyPages[sourceLanguage]?.length) {
      return;
    }
    /** @type {Page[]} */
    const pages = await pageApi.fetchNearbyPages(sourceLanguage);

    commit("addNearbyPages", { language: sourceLanguage, pages });
  }
};
