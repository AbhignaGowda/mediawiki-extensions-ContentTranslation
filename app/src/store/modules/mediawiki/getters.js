import MTProviderGroup from "../../../wiki/mw/models/mtProviderGroup";

export default {
  /**
   * @param {Object} state
   * @return {function(string, string): Page}
   */
  getPage: state => (language, title) =>
    state.pages.find(
      page =>
        page.language === language &&
        (page.title === title || page.alias === title)
    ),

  getLanguageTitleGroup: state => (language, title) =>
    state.languageTitleGroups.find(group =>
      group.titles.find(
        groupTitle => groupTitle.lang === language && groupTitle.title === title
      )
    ),

  getLanguageTitleGroupByWikidataId: state => wikidataId =>
    state.languageTitleGroups.find(group => group.wikidataId === wikidataId),

  getTitleByLanguageForGroup: (state, getters) => (wikidataId, language) =>
    (getters.getLanguageTitleGroupByWikidataId(wikidataId)?.titles || []).find(
      title => title.lang === language
    )?.title,

  /**
   * Get the language object for the given language code
   * @param {String} languageCode
   * @returns {Language}
   */
  getLanguage: state => languageCode =>
    state.languages.find(language => language.code === languageCode),

  getPageSection: state => (page, sectionTitle) => {
    return (page?.sections || []).find(
      section => section.originalTitle === sectionTitle
    );
  },

  /**
   * Get MTProviderGroup for the given language pair
   * @param {String} sourceLanguage
   * @param {String} targetLanguage
   * @returns {String[]} - Array of supported providers
   */
  getSupportedMTProviders: state => (sourceLanguage, targetLanguage) =>
    state.supportedMTProviderGroups.find(
      mtProviderGroup =>
        mtProviderGroup.sourceLanguage === sourceLanguage &&
        mtProviderGroup.targetLanguage === targetLanguage
    )?.providers || [],

  isValidProviderForTranslation: (state, getters) => (
    sourceLanguage,
    targetLanguage,
    provider
  ) =>
    getters
      .getSupportedMTProviders(sourceLanguage, targetLanguage)
      .includes(provider) &&
    provider !== MTProviderGroup.EMPTY_TEXT_PROVIDER_KEY,
  /**
   * Get recently edited cx translations by current user if any,
   * for the current language pair.
   *
   * @param {Object} state
   * @param {Object} rootGetters
   * @param {Object} getters
   * @param {Object} rootState
   * @return {Page[]}
   */
  getRecentlyEditedPages: (state, getters, rootState, rootGetters) => {
    const sourceLanguage = rootState.application.sourceLanguage;
    const targetLanguage = rootState.application.targetLanguage;
    /** @type Translation[] */
    const translations = rootGetters[
      "translator/getAllTranslationsForLanguagePair"
    ](sourceLanguage, targetLanguage);
    const translationsSlice = translations.slice(
      0,
      rootState.suggestions.maxRecentlyEditedSuggestions
    );
    return translationsSlice.map(translation =>
      getters.getPage(sourceLanguage, translation.sourceTitle)
    );
  }
};
