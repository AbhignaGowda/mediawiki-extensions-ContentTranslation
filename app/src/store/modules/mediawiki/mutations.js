import Vue from "vue";

export default {
  addPage(state, page) {
    state.pages.push(page);
  },

  addLanguageTitleGroup(state, group) {
    state.languageTitleGroups.push(group);
  },

  setLanguages(state, languages) {
    state.languages = languages;
  },

  setSupportedLanguageCodes(state, languageCodes) {
    state.supportedLanguageCodes = languageCodes;
  },

  /**
   * @param mtProviderGroup
   */
  addMtProviderGroup(state, mtProviderGroup) {
    state.supportedMTProviderGroups.push(mtProviderGroup);
  },

  setPageSections(state, { page, sections }) {
    Vue.set(page, "sections", sections);
  },

  setLanguagesRequested(state, value) {
    state.languagesRequested = value;
  },

  setSupportedLanguageCodesRequested(state, value) {
    state.supportedLanguageCodesRequested = value;
  }
};
