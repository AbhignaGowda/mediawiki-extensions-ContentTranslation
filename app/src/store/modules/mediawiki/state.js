export default {
  /** @type {Page[]} */
  pages: [],
  /** @type {Language[]} */
  languages: [],
  /** @type {Boolean} */
  languagesRequested: false,
  languageTitleGroups: [],
  supportedLanguageCodes: [],
  /** @type {Boolean} */
  supportedLanguageCodesRequested: false,
  supportedMTProviderGroups: [],
  /**
   * Stores nearby pages to be used as suggestions for section translation
   * Format: {{ en: Page[], es: Page[], ... }}
   * @type {Object}
   */
  nearbyPages: {}
};
