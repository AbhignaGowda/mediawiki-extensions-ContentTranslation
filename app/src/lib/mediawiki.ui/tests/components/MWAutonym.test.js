import Vuex from "vuex";
import { shallowMount, createLocalVue } from "@vue/test-utils";
import MwAutonym from "../../components/MWAutonym";
import mediawiki from "@/store/modules/mediawiki";

describe("MwAutonym.vue", () => {
  let store,
    localVue,
    storeConfig = { modules: { mediawiki } };

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(Vuex);
    store = new Vuex.Store(storeConfig);
  });

  it("renders the autonym for the passed language", () => {
    const lang = "ml";
    const autonym = "മലയാളം";
    const dir = "ltr";
    const languageInfo = {
      ml: {
        autonym,
        dir
      }
    };
    expect(Object.keys(store.state.mediawiki.languageInfo).length).toBe(0);
    store.commit("mediawiki/setLanguageInfo", languageInfo);
    const wrapper = shallowMount(MwAutonym, {
      propsData: { lang },
      store,
      localVue
    });
    expect(wrapper.find("span").text()).toMatch("മലയാളം");
    expect(wrapper.find("span").attributes("lang")).toBe(lang);
    expect(wrapper.find("span").attributes("dir")).toBe(dir);
    expect(wrapper.element).toMatchSnapshot();
  });
});
