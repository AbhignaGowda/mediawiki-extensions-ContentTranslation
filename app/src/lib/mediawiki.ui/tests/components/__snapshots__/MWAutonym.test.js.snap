// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MwAutonym.vue renders the autonym for the passed language 1`] = `
<span
  class="mw-ui-autonym"
  dir="ltr"
  lang="ml"
>
  മലയാളം
</span>
`;
